### PR TITLE
logger 'async' sending mechanism removed

### PR DESCRIPTION
**Pull request by guliashvili**
Recently several issues have been open about the logger. #4608 #4983

In an effort to simplify logging mechanism, removing the async logging which used additional threads for buffering and sending the logs.
async logging had several drawbacks.

- Spawning additional threads make osquery use more system resources.
- It's still not the big change in behavior. The next thing, that would happen after spawning the thread was the WaitForSinks to be called, which would wait for the thread to finish.

Feel free to consult with the GLOG implementation/comments

https://github.com/google/glog/blob/0472b91c5defdf90cff7292e3bf7bd86770a9a0a/src/glog/logging.h.in#L1407
https://github.com/google/glog/blob/master/src/logging.cc#L1471
https://github.com/google/glog/blob/master/src/logging.cc#L1349
https://github.com/google/glog/blob/master/src/logging.cc#L795

### CODE REVIEW

**Comment by guliashvili** [comment]
ok to test

File: osquery/dispatcher/scheduler.cpp [hunk 1]
```diff
@@ -211,7 +211,7 @@ void SchedulerRunner::start() {
 
     // GLog is not re-entrant, so logs must be flushed in a dedicated thread.
     if ((i % 3) == 0) {
-      relayStatusLogs(true);
```
Comment (fmanco) [inline osquery/dispatcher/scheduler.cpp]: This is called on the main scheduler thread (it actually should be moved to and independent thread), so making `relayStatusLogs` synchronous has the potential of delaying this thread.

Comment (guliashvili) [reply osquery/dispatcher/scheduler.cpp]: Nice catch.  I blindly changed it to fit the new interface. This function does not need to be called here at all, removing it.

**Comment by guliashvili** [comment]
@fmanco 
Thanks for the review. Your comment was legit, as relayStatusLogs was called from the thread not actually logging. However, it's not needed, so removing it. In other cases, my argument is valid.

**Comment by guliashvili** [comment]
@fmanco 
It's not needed because, if GLOG logs something it will also call WaitTillSent directly after logging. WaitTillSent does nothing but calling relayStatusLogs. My argument of no increased sync time hold true, because in the old code, WaitTillSent was just waiting for the thread to finish, now it does the job. But time being in the WaitTillSent stays same.

**Comment by guliashvili** [comment]
@fmanco 
That's true, things will be synchronous now.
Problem with the logging is that it's too complicated and tries to squeeze out the performance using multithreading.

To my understanding, making things fast by saving CPU cycles is good, however making things fast by parallelization, in case of the osquery, is not that good. Because of 1. The parallelized code in total always wastes from CPU cycles than it's single threaded counterpart 2. osquery will be prone to larger CPU usage spikes.

During the whole lifecycle of the log, it is buffered/batched multiple times.
1) inside GLOG
2) Was batched in BufferedLogSink
3) Is batched in BufferedLogForwarder (Most of the plugins implement it and if they don't they should)

The problem with BufferedLogSink that stands it out from other batching mechanisms is exactly multithreading.

It's also worth mentioning that when plugin implements BufferedLogForwarder(for example TLS) logging thread will just wait until the log is saved in the database and not until the log is delivered to the endpoint.

**Pull request closed without merging**